### ISSUE TITLE
Clarification questions

### ISSUE THREAD

**Issue by wasiahmad** [facebookresearch/CodeGen#23]
Hi, I have a few questions regarding TransCoder's training data and optimization setting.

1. From the paper, it is clear that TransCoder is trained using Standalone functions during the DAE+BT training stage. But is TransCoder only trained using Standalone functions in the MLM stage too?
2. During the MLM stage, only the encoder part of TransCoder is pre-trained, right?
3. For the MLM pre-training, `max_epoch` and `epoch_size` are set to 100k. If I understand correctly, `epoch_size` basically refers to the number of instances used in each epoch. Is it correct? Also, for MLM pre-training, the following are set:

```
--validation_metrics _valid_mlm_ppl \
--stopping_criterion '_valid_mlm_ppl,10' 
```

So, I am assuming TransCoder pre-training is stopped based on the `stopping_criterion`. Before, the MLM pre-training was stopped, how many optimization steps were executed?

4. Unlike the MLM pre-training stage, for the DAE+BT stage training, there is no `stopping_criterion` is set. And the `epoch_size` was set to 50000 and the `max_epoch` was set to 10000000. So, when the training stops? How many optimization steps were executed during this stage?

**Comment by brozi** [comment]
1. We train on whole files for the MLM stage
2. Yes but we reload the encoder pre-trained with MLM for the decoder when training TransCoder
3. It stopped at 557 epochs on 32 GPUs for the model we release in this repo. 
4. We stopped the training when it stopped improving and selected the best model using the validation set or stopped it after two days of training for the translation models trained from DOBF. The TransCoder_model_1.pth corresponds to the 30th epochs and TransCoder_model_2.pth to the 33rd.

**Comment by wasiahmad** [comment]
@brozi one more question, MLM pre-training is done with a batch size 32. Is it per-GPU-batch-size or total batch size? If it is the total batch size, then I am curious to know why did you need 32 V100 GPUs? Or, actually, the total batch size is (32 x 32) = 1024?

**Comment by brozi** [comment]
It's per GPU so yes the actual batch size is 32 * 32

**Comment by wasiahmad** [comment]
One more question for my understanding. You said, `The TransCoder_model_1.pth corresponds to the 30th epochs and TransCoder_model_2.pth to the 33rd (and the epoch_size was 30k instead of 50k)`. I want to understand this part clearly.

Does epoch size 30k mean one epoch composed of 30,000 standalone functions? If yes, then stopping after 30 epochs mean, the model was trained with (30,000 x 30) = 900k functions? (I doubt this is the case)

**OR**, epoch size 30k means 30k updates (each update is associated with 1024 instances, which is the batch size)? In this case, 30.72M functions are used in each epoch. Stopping after 30 epochs would mean the model is trained with 921.6M functions.

From the TransCoder paper, it seems like there are 60M Java, 60M C++, and 109M Python Standalone functions (15%, 50%, 50% of the total Java, C++, and Python functions). So, in total there are (approx.) 229M Standalone functions. So, if the above-mentioned calculation is correct, TransCoder went through the entire collection 4 times (roughly).

**Comment by brozi** [comment]
The epoch size is supposed to be the number of samples you train on. 
However there is something that's a bit confusing in our code. We always use increase the sentences counter by the batch_size parameter when training but it's not always the actual batch size.
- for stream datasets (e.g. for MLM) we use the batch parameter so epoch_size is really the number of samples (per GPU so you need to multiply by 32)
- for DAE and BT we fix the number of tokens per batch instead of the epoch size to avoid OOMs due to batches of varying sizes so the batch size is not constant. In that case epoch_size is num_updates/32 (32 is the default batch size parameter), which doesn't really cause any issues but is quite confusing. 
I'll think about changing the behaviour of this parameter to have something more coherent in a way that minimizes the changes people will need to make to the parameter.
